replace_with/src/lib.rs

```diff
@@ -0,0 +1,8 @@
+use std::ptr;
+
+/// Replaces `*t` with `f` applied to the original `*t`,
+pub fn replace_with<T, F: FnOnce(T) -> T>(t: &mut T, f: F) {
+    unsafe {
+        *t = f(ptr::read(t));
```

Comment:
It turns out that assigning to `*t` where `t` is an `&mut` (or a raw pointer, for that matter) causes the original contents to be dropped. What you actually want here is `ptr::write`, which does the assignment without any drop.

This fails one of the provided tests for the above reason.